File: src/test/java/de/fluxparticle/animation/bubblesort/BubbleSortBox.java
package de.fluxparticle.animation.bubblesort;

import de.fluxparticle.animation.AnimationQueue;
import de.fluxparticle.animation.object.ElementObject;
import de.fluxparticle.animation.object.ElementRectangle;
import de.fluxparticle.animation.point.DynamicPoint2D;
import javafx.beans.binding.DoubleExpression;
import javafx.geometry.BoundingBox;
import javafx.geometry.Bounds;
import javafx.scene.paint.Color;

/**
 * Created by sreinck on 09.02.16.
 */
public class BubbleSortBox {

    private static final int WIDTH = ElementObject.OPERAND_WIDTH;

    private static final int HEIGHT = ElementObject.HEIGHT;

    private static final int SPACING = 10;

    private final int[] array = { 13, 4, 3, 21, 1 };

    private final ElementObject[] elementObjects = new ElementObject[array.length];

    private final AnimationQueue animationQueue;

    private ElementRectangle rectangleCompare;

    private ElementRectangle rectangleReady;

    public BubbleSortBox(AnimationQueue animationQueue) {
        this.animationQueue = animationQueue;

        animationQueue.with((aq, t) -> {
            for (int i = 0; i < array.length; i++) {
                int value = array[i];

                elementObjects[i] = new ElementObject(t, true, true, WIDTH, Integer.toString(value));

                DynamicPoint2D translate = center(i);
                elementObjects[i].setTranslate(translate);

                aq.fadeIn(0, elementObjects[i]);
            }
        });
    }

    public Bounds getBounds() {
        return new BoundingBox(0, 0, SPACING + array.length * (WIDTH + SPACING), 2 * SPACING + HEIGHT);
    }

    public int size() {
        return array.length;
    }

    public boolean inOrder(int index) {
        boolean result = array[index] < array[index + 1];

        animationQueue.with((aq, t) -> {

            Bounds bs = surround(index, 2);

            Color colorCompare = Color.BLUE;
            rectangleCompare = createOrMove(aq, t, bs, colorCompare, rectangleCompare);

            Color c = result ? Color.GREEN : Color.RED;
            aq.changeColor(1, rectangleCompare, c);
        });

        return result;
    }

    public void swap(int index) {
        int next = index + 1;
        {
            int tmp = array[index];
            array[index] = array[next];
            array[next] = tmp;
        }

        animationQueue.with((aq, t) -> {
            aq.move(0, elementObjects[index], center(next), false, null, true);
            aq.move(0, elementObjects[next], center(index), false, null, true);

            ElementObject tmp = elementObjects[index];
            elementObjects[index] = elementObjects[next];
            elementObjects[next] = tmp;
        });
    }

    public void ready(int index) {
        int range = array.length - index;

        animationQueue.with((aq, t) -> {
            if (rectangleCompare != null) {
                aq.fadeOut(-1, rectangleCompare);
                rectangleCompare = null;
            }

            Bounds bs = surround(index, range);

            Color colorReady = Color.GREEN;
            rectangleReady = createOrMove(aq, t, bs, colorReady, rectangleReady);
        });
    }

    private static ElementRectangle createOrMove(AnimationQueue aq, DoubleExpression t, Bounds bs, Color colorReady, ElementRectangle rectangle) {
        if (rectangle == null) {
            rectangle = new ElementRectangle(t, bs, colorReady);
            aq.fadeIn(0, rectangle);
        } else {
            aq.adjustAreaToBounds(0, rectangle, bs);
            aq.changeColor(0, rectangle, colorReady);
        }

        return rectangle;
    }

    private static DynamicPoint2D center(int index) {
        return new DynamicPoint2D(SPACING + WIDTH / 2 + index * (WIDTH + SPACING), SPACING + HEIGHT / 2);
    }

    private static BoundingBox surround(int index, int range) {
        return new BoundingBox(SPACING / 2 + index * (WIDTH + SPACING), SPACING / 2, range * (WIDTH + SPACING), HEIGHT + SPACING);
    }

}

File: jvm/src/test/kotlin/de/fluxparticle/animation/bubblesort/BubbleSortBox.kt
package de.fluxparticle.animation.bubblesort

import de.fluxparticle.animation.AnimationQueue
import de.fluxparticle.animation.elementobject.ElementObject
import de.fluxparticle.animation.elementobject.ElementRectangle
import de.fluxparticle.animation.point.DynamicPoint2D
import de.fluxparticle.animation.signal.Signal
import de.fluxparticle.animation.util.Bounds
import de.fluxparticle.animation.util.Color
import de.fluxparticle.animation.value.Value
import java.util.concurrent.atomic.AtomicReference

/**
 * Created by sreinck on 09.02.16.
 */
class BubbleSortBox(private val animationQueue: AnimationQueue) {

    private val array = intArrayOf(13, 4, 3, 21, 1)

    private val elementObjects = Array<ElementObject>(array.size) { i ->
        ElementObject(animationQueue.time, true, true, WIDTH, array[i].toString()).apply {
            this.translate = Signal(animationQueue.time, center(i))
        }
    }

    private var rectangleCompare = AtomicReference<ElementRectangle>()

    private var rectangleReady = AtomicReference<ElementRectangle>()

    val bounds: Bounds
        get() = Bounds(0.0, 0.0, (SPACING + array.size * (WIDTH + SPACING)).toDouble(), (2 * SPACING + HEIGHT).toDouble())

    init {
        animationQueue.with { aq, t ->
            for (i in array.indices) {
                aq.fadeIn(0, elementObjects[i])
            }
        }
    }

    fun size(): Int {
        return array.size
    }

    fun inOrder(index: Int): Boolean {
        val result = array[index] < array[index + 1]

        animationQueue.with { aq, t ->

            val bs = surround(index, 2)

            val colorCompare = Color.BLUE
            rectangleCompare.createOrMove(aq, t, bs, colorCompare)

            val c = if (result) Color.GREEN else Color.RED
            aq.changeColor(1, rectangleCompare.get(), c)
        }

        return result
    }

    fun swap(index: Int) {
        val next = index + 1
        run {
            val tmp = array[index]
            array[index] = array[next]
            array[next] = tmp
        }

        animationQueue.with { aq, t ->
            aq.move(0, elementObjects[index], center(next), false, null, true)
            aq.move(0, elementObjects[next], center(index), false, null, true)

            val tmp = elementObjects[index]
            elementObjects[index] = elementObjects[next]
            elementObjects[next] = tmp
        }
    }

    fun ready(index: Int) {
        val range = array.size - index

        animationQueue.with { aq, t ->
            val rectangle = rectangleCompare.get()
            if (rectangle != null) {
                aq.fadeOut(-1, rectangle)
                rectangleCompare.set(null)
            }

            val bs = surround(index, range)

            val colorReady = Color.GREEN
            rectangleReady.createOrMove(aq, t, bs, colorReady)
        }
    }

    companion object {

        private val WIDTH = ElementObject.OPERAND_WIDTH

        private val HEIGHT = ElementObject.HEIGHT

        private val SPACING = 10

        private fun center(index: Int): DynamicPoint2D {
            return DynamicPoint2D((SPACING + WIDTH / 2 + index * (WIDTH + SPACING)).toDouble(), (SPACING + HEIGHT / 2).toDouble())
        }

        private fun surround(index: Int, range: Int): Bounds {
            return Bounds((SPACING / 2 + index * (WIDTH + SPACING)).toDouble(), (SPACING / 2).toDouble(), (range * (WIDTH + SPACING)).toDouble(), (HEIGHT + SPACING).toDouble())
        }
    }

}

private fun AtomicReference<ElementRectangle>.createOrMove(aq: AnimationQueue, t: Value<Double>, bs: Bounds, colorReady: Color) {
    val rectangle = this.get()
    if (rectangle == null) {
        val newRectangle = ElementRectangle(t, bs, colorReady)
        aq.fadeIn(0, newRectangle)
        set(newRectangle)
    } else {
        aq.adjustAreaToBounds(0, rectangle, bs)
        aq.changeColor(0, rectangle, colorReady)
    }
}
